File: Trees/Height_Of_K-Ary_Tree/solutions/optimal_solution.cpp
#include<bits/stdc++.h>

using namespace std;

//--------------------------------START------------------------------------

const int MAX_N = 100000;

struct node
{
	//int val;														// To find height of tree, value stored in nodes does not matter. So in input also we are not given this field. 
	vector<node*> children;											// Address of children nodes. 
	node()
	{
		children.clear();
	}
};

int find_max_height(node* root)
{
	if (root->children.size() == 0)									// Handle base case when root is a leaf node.
	{
		return 0;
	}
	int h = 0;
	for (int i = 0; i < root->children.size(); i++)
	{
		h = max(h, find_max_height(root->children[i]));				// Find height of each children and store the maximum height of children. 
	}
	return h + 1;			
}

unordered_map<int, node*> address; 									// To build k-ary tree use this to speed up the insertion process. 

int find_height(int k, vector<int> from, vector<int> to)
{
	int N = from.size() + 1;
	address.clear();												// Clear the global variable. 
	for (int i = 1; i <= N; i++)
	{
		address[i] = new node();									// Create N nodes. 
	}
	for (int i = 0; i < N - 1; i++)								  
	{
		address[from[i]]->children.push_back(address[to[i]]);		// Link the nodes. (Build the k-ary tree.)
	}
	return find_max_height(address[1]);									
}

//--------------------------------STOP------------------------------------

int main()
{
	//freopen("..//test_cases//sample_test_cases_input.txt", "r", stdin);
	//freopen("..//test_cases//sample_test_cases_output.txt", "w", stdout);
	//freopen("..//test_cases//handmade_test_cases_input.txt", "r", stdin);
	//freopen("..//test_cases//handmade_test_cases_output.txt", "w", stdout);
	//freopen("..//test_cases//small_test_cases_input.txt", "r", stdin);
	//freopen("..//test_cases//small_test_cases_output.txt", "w", stdout);
	freopen("..//test_cases//big_test_cases_input.txt", "r", stdin);
	freopen("..//test_cases//big_test_cases_output.txt", "w", stdout);
	//freopen("..//test_cases//ignore.txt", "w", stdout);

	int test_cases;
	cin >> test_cases;
	while (test_cases--)
	{
		int k;
		cin >> k;

		int edges;
		cin >> edges;
		
		int N = edges + 1;
		assert(1 <= N);
		assert(N <= MAX_N);
		assert(0 <= k);
		assert(k <= N - 1);
		
		vector<int> from(edges, 0);
		for (int i = 0; i < edges; i++)
		{
			cin >> from[i];
			assert(1 <= from[i]);
			assert(from[i] <= N);
		}
		
		cin >> edges;
		assert(edges == N - 1);
		
		vector<int> to(edges, 0);
		for (int i = 0; i < edges; i++)
		{
			cin >> to[i];
			assert(1 <= to[i]);
			assert(to[i] <= N);
		}

		unordered_map<int, int> no_of_in_edges;
		for (int i = 0; i < edges; i++)
		{
			no_of_in_edges[to[i]]++;
		} 
		assert(no_of_in_edges.size() == N - 1);

		for (auto it = no_of_in_edges.begin(); it != no_of_in_edges.end(); it++)
		{
			assert(it->second == 1);
		}

		int h = find_height(k, from, to);
		assert(0 <= h);
		assert(h <= N - 1);
		cout << h << endl;
	}

	return 0;
}

File: Trees/Height_Of_K-Ary_Tree/solutions/optimal_solution.js
process.stdin.resume();
process.stdin.setEncoding('utf-8');

var __input_stdin = "";
var __input_stdin_array = "";
var __input_currentline = 0;

process.stdin.on('data', function(data) {
    __input_stdin += data;
});

function TreeNode()
{
    //this.val = 0;														// To find height of tree, value stored in nodes does not matter. So in input also we are not given this field. 
    this.children = [];
}

//--------------------------------START------------------------------------

/*
 * Complete the function below.
 */

/*
    For your reference:
    
    function TreeNode()
	{
	    this.children = [];
	}
*/

function find_height(k, root) 
{
    if (root.children.length == 0)									// Handle base case when root is a leaf node.
    {
        return 0;
    }
    var h = 0;
    for (var i = 0; i < root.children.length; i++)
    {
        var temp = root.children[i];
        h = Math.max(h, find_height(k, temp));				       // Find height of each children and store the maximum height of children. 
    }
    return h + 1;	
}

//--------------------------------STOP------------------------------------

var address = {};

function build_tree(from, to) {
    var N = from.length + 1;
    address = {};												    // Clear the global variable. 
    for (var i = 1; i <= N; i++)
    {
        address[i] = new TreeNode();								// Create N nodes. 
    }
    for (var i = 0; i < N - 1; i++)								  
    {
        address[from[i]].children.push(address[to[i]]);		        // Link the nodes. (Build the k-ary tree.)
    }
    return address[1];	
}

var fs = require('fs');
var wstream = fs.createWriteStream(process.env.OUTPUT_PATH);

process.stdin.on('end', function() {
    __input_stdin_array = __input_stdin.split("\n");
    var k;
    var k = parseInt(__input_stdin_array[__input_currentline].trim(), 10);
    __input_currentline += 1;

    var from;
    var from_size = 0;
    from_size = parseInt(__input_stdin_array[__input_currentline].trim(), 10);
    __input_currentline += 1;

    var from = [];
    var from_item;
    for (var from_i = 0; from_i < from_size; from_i++) {
        var from_item = parseInt(__input_stdin_array[__input_currentline].trim(), 10);
        __input_currentline += 1;
        from.push(from_item);
    }

    var to;
    var to_size = 0;
    to_size = parseInt(__input_stdin_array[__input_currentline].trim(), 10);
    __input_currentline += 1;

    var to = [];
    var to_item;
    for (var to_i = 0; to_i < to_size; to_i++) {
        var to_item = parseInt(__input_stdin_array[__input_currentline].trim(), 10);
        __input_currentline += 1;
        to.push(to_item);
    }

    var root = build_tree(from, to);

    res = find_height(k, root);
    wstream.write(res + "\n");

    wstream.end();
});